### ISSUE TITLE
Missing = in const declaration

### ISSUE THREAD

**Issue by WebReflection** [cherow/cherow#42]
The following is perfectly valid ES code.
```js
const fn = c => {
  let a, b; // this fails
  if (c) {
    a = c;
    b = 2;
  } else {
    a = 2;
    b = 3;
  }
  return a + b;
};
```
This parser wants every `let` declaration to be assigned, misreading them as constants.

`let` variables are not constants.

Thanks for fixing this.

**Comment by WebReflection** [comment]
P.S. exact same goes for `var` ... basically _cherow_ doesn't like single variable declaration which makes it unusable with standard JS code.

**Comment by WebReflection** [comment]
**update** it's only in scope.

The following fails:
```js
const fn = c => {
  let a;
  if (c) a = 1;
  else a = 0;
  return a;
};
```
but this doesn't
```js
let a;
if (c) a = 1;
else a = 0;
```

**Comment by WebReflection** [comment]
PR landed. Yet the `var a, b;` case is not covered but the logic for `var` is different.

If you have any idea how to reliably solve `var` please let me know so I can update the PR.

Thanks.

**Comment by KFlash** [comment]
Hi. Sorry late reply. And thanks for finding this. `var` should be easy. `const` and `let` have a common mask `lexical` so separate them from `var`. So this check `if !(context & Context.Lexical)`  are only valid for `var`. I

**Comment by WebReflection** [comment]
I hope you didn't choke while writing that ...

anyway `var` fails just like `let`.

```js
const fn = c => {
  var a;
};
```

fails and it entered in the exact same `context & Context.Lexical` condition

**Comment by WebReflection** [comment]
if you see my PR test:
https://github.com/cherow/cherow/pull/43/commits/4b8ec601f582489e87cfab20995850eaf5066a01#diff-92c09fcab6c9b3abd2395789365f2172

you can verify yourself if you write the following the test fails because `var` reaches that part:
```js
it('should not fail with missed, in scope, let declaration', () => {
    expect(() => { parseScript('const f = () => { let v; var z; };'); }).to.not.throw();
});
```

**Comment by KFlash** [comment]
@WebReflection I'm on my mobile and on my way home, so wrote it while driving. This issue seem to happen during the refactoring. See the old solution for it [here](https://github.com/cherow/cherow/blob/25c0ed6f95a69ce0d1033e89831e018c74afb16a/src/parser.ts#L3101)

**Comment by WebReflection** [comment]
yeah, please don't text & drive
I can survive with this not working 'till you're safe home

**Comment by KFlash** [comment]
@WebReflection I'm home soon! Anyway. Your issue seem to be related to body of the arrow function. So somewhere down the road the mask isn't unset. As a workaround for now you need to unset this in the function body. Find this `private parseFunctionBody(context: Context): ESTree.BlockStatement {}` and after the position you can add `context &= ~Context.Lexical`. I think that should solve it.
If the issue got solved, need to investigate where to correctly unset the mask

**Comment by KFlash** [comment]
I stopped the car and was thinking about it. Just backup the context mask before entering the function body, unset it and set the mask back on when done. Something like this.

```
const savedLexical = (context & Context.Lexical) !== 0;
context &= ~Context.Lexical
//... enter the function body
if (savedLexical) context |= Context.Lexical;
```
Will check when I got back home

**Comment by KFlash** [comment]
@WebReflection Just found a McDonalds with free WiFI. So this will be fixed within 10 minutes.

**Comment by KFlash** [comment]
Verify if this solved your problem. Now I need to get back home :)

**Comment by WebReflection** [comment]
Thanks, I'll test after dinner.

If it works, should I close my PR? Is there any test to validate both `let` and `var` ?

I think this would be the bare minimum:
```js
it('should not fail with missed, in scope, let or var declaration', () => {
    expect(() => { parseScript('const f = () => { let v; var z; };'); }).to.not.throw();
});
```

I've used this in `let.js` to validate my tests

**Comment by KFlash** [comment]
In my quick fix I added in your example tests and they pass. I'm home very soon now and will push a few more tests. 
Regarding the PR. I think you can close it. The issue was that it didn't unset the context masks correctly in function bodies. However. I haven't tested for `BlockStatement`, but will do.

**Comment by KFlash** [comment]
Just got home and pushed a permanent fix for this. I may have overlooked something... Not sure.

**Comment by WebReflection** [comment]
> I may have overlooked something... Not sure.

I have the following issue now:

`Export declarations may only appear at top level of a module`

Something ain't right ...

**Comment by WebReflection** [comment]
TL;DR [this](https://github.com/WebReflection/majinbuu) is the module I want/need to parse and make a CommonJS module via [ascjs](https://github.com/WebReflection/ascjs) parser based on _cherow_.

```js
/*! Copyright (c) 2017, Andrea Giammarchi, @WebReflection */

// grid operations
const DELETE = 'del';
const INSERT = 'ins';
const SUBSTITUTE = 'sub';

// typed Array
const TypedArray = typeof Int32Array === 'function' ? Int32Array : Array;

const majinbuu = (from, to, MAX_SIZE) => {

  const fromLength = from.length;
  const toLength = to.length;
  const TOO_MANY = (MAX_SIZE || Infinity) < Math.sqrt((fromLength || 1) * (toLength || 1));

  if (fromLength < 1 || TOO_MANY) {
    if (toLength || TOO_MANY) {
      from.splice.apply(from, [0, fromLength].concat(to));
    }
    return;
  }
  if (toLength < 1) {
    from.splice(0);
    return;
  }
  performOperations(
    from,
    getOperations(from, to, levenstein(from, to))
  );
}; 

// given an object that would like to intercept
// all splice operations performed through a list,
// wraps the list.splice method to delegate such object
// and it puts back original splice right before every invocation.
// Note: do not use the same list in two different aura
const aura = (splicer, list) => {
  const splice = list.splice;
  const $splice = (...args) => {
    list.splice = splice;
    const result = splicer.splice(...args);
    list.splice = $splice;
    return result;
  };
  list.splice = $splice;
  return list;
};

// Helpers - - - - - - - - - - - - - - - - - - - - - -

// originally readapted from:
// http://webreflection.blogspot.co.uk/2009/02/levenshtein-algorithm-revisited-25.html
// then rewritten in C for Emscripten (see levenstein.c)
// then "screw you ASM" due no much gain but very bloated code
const levenstein = (from, to) => {
  const fromLength = from.length + 1;
  const toLength = to.length + 1;
  const size = fromLength * toLength;
  const grid = new TypedArray(size);
  let x = 0;
  let y = 0;
  let X = 0;
  let Y = 0;
  let crow = 0;
  let prow = 0;
  let del, ins, sub;
  grid[0] = 0;
  while (++x < toLength) grid[x] = x;
  while (++y < fromLength) {
    X = x = 0;
    prow = crow;
    crow = y * toLength;
    grid[crow + x] = y;
    while (++x < toLength) {
      del = grid[prow + x] + 1;
      ins = grid[crow + X] + 1;
      sub = grid[prow + X] + (from[Y] == to[X] ? 0 : 1);
      grid[crow + x] = del < ins ?
                        (del < sub ?
                          del : sub) :
                        (ins < sub ?
                          ins : sub);
      ++X;
    };
    Y = y;
  }
  return grid;
};

// add operations (in reversed order)
const addOperation = (list, type, x, y, count, items) => {
  list.unshift({type, x, y, count, items});
};

// walk the Levenshtein grid bottom -> up
const getOperations = (Y, X, grid) => {
  const list = [];
  const YL = Y.length + 1;
  const XL = X.length + 1;
  let y = YL - 1;
  let x = XL - 1;
  let cell,
      top, left, diagonal,
      crow, prow;
  while (x && y) {
    crow = y * XL + x;
    prow = crow - XL;
    cell = grid[crow];
    top = grid[prow];
    left = grid[crow - 1];
    diagonal = grid[prow - 1];
    if (diagonal <= left && diagonal <= top && diagonal <= cell) {
      x--;
      y--;
      if (diagonal < cell) {
        addOperation(list, SUBSTITUTE, x, y, 1, [X[x]]);
      }
    }
    else if (left <= top && left <= cell) {
      x--;
      addOperation(list, INSERT, x, y, 0, [X[x]]);
    }
    else {
      y--;
      addOperation(list, DELETE, x, y, 1, []);
    }
  }
  while (x--) {
    addOperation(list, INSERT, x, y, 0, [X[x]]);
  }
  while (y--) {
    addOperation(list, DELETE, x, y, 1, []);
  }
  return list;
};

/* grouped operations */
const performOperations = (target, operations) => {
  const length = operations.length;
  let diff = 0;
  let i = 1;
  let curr, prev, op;
  if (length) {
    op = (prev = operations[0]);
    while (i < length) {
      curr = operations[i++];
      if (prev.type === curr.type && (curr.x - prev.x) <= 1 && (curr.y - prev.y) <= 1) {
        op.count += curr.count;
        op.items = op.items.concat(curr.items);
      } else {
        target.splice.apply(target, [op.y + diff, op.count].concat(op.items));
        diff += op.type === INSERT ?
          op.items.length : (op.type === DELETE ?
            -op.count : 0);
        op = curr;
      }
      prev = curr;
    }
    target.splice.apply(target, [op.y + diff, op.count].concat(op.items));
  }
};

// module
majinbuu.aura = aura;
export {majinbuu, aura};
export default majinbuu;
```

This module works perfectly fie with ES2015+ compatible browsers.

Right now this module fails to be parsed and I'm not sure I understand why.

**Comment by KFlash** [comment]
Well. I like challenges :) Give me a few minutes ...

**Comment by KFlash** [comment]
@WebReflection Could you try to parse your code again?

**Comment by WebReflection** [comment]
not sure how to test. I don't see tests in the latest commit and I am using your published npm versions through ascjs project.

When you think you've fixed this, please publish to npm and I'll confirm.

Meanwhile, I'll close this bug because the current one is totally different.

**A question** though, how much code, beside cherow own tests, have been parsed and validated in there? These two issues were kinda trivial, regular, ES2015+ modules code and I desperately need a robust, and why not fast, solution.

I don't want to sound rude, and I'd love to help here as much as possible, but these two bugs made me think maybe, for my current needs, cherow is not the best pick in terms of standard ES2015+ parsing?

**Comment by KFlash** [comment]
Well. It's your decision. And ATM I got big time issue publishing to NPM, but the issues are solved.
Regarding your other concerns it's tested on various libraries, but I will spend a day not to go through all GH code I can find and fix the bugs I may find.

**Comment by KFlash** [comment]
@WebReflection I will ping you again when I validated Cherow aginst all online libs. And NPM version can't be added due to NPM issue

```js
npm ERR! registry error parsing json
npm ERR! registry error parsing json
npm ERR! registry error parsing json
npm ERR! publish Failed PUT 503
npm ERR! Unexpected token < in JSON at position 1
npm ERR!
npm ERR! <?xml version="1.0" encoding="utf-8"?>
npm ERR! <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
npm ERR!  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
npm ERR! <html>
npm ERR!   <head>
npm ERR!     <title>503 client read error</title>
npm ERR!   </head>
npm ERR!   <body>
npm ERR!     <h1>Error 503 client read error</h1>
npm ERR!     <p>client read error</p>
npm ERR!     <h3>Guru Mediation:</h3>
npm ERR!     <p>Details: cache-sin18024-SIN 1508289847 1180018762</p>
npm ERR!     <hr>
npm ERR!     <p>Varnish cache server</p>
npm ERR!   </body>
npm ERR! </html>
npm ERR!

npm ERR! A complete log of this run can be found in:

```

**Comment by KFlash** [comment]
@WebReflection Updated latest to NPM.

I have now also parsed all your repos now and no bugs found.  However you got an issue in a few of your libraries, and for  this [Import()](https://github.com/WebReflection/import.js) repo of yours you would need to activate the `next option` during parsing. But parses without any issues!

[this file](https://raw.githubusercontent.com/WebReflection/dns.js.org/master/cnames_active.js) for the [dns.js.org](https://github.com/WebReflection/dns.js.org) contain invalid JS. Also confirmed with other parsers through AST Explorer. It complaining about a invalid period on line 244 in all parsers. Including Cherow.

I will keep parsing through GH during the day to catch the bugs I may find. I agree that the bug you found was trivial, but same thing can and also have happend on other parsers.

**Comment by KFlash** [comment]
Just add to the record that I know ran through a lot of libraries like `React`, `Angular`, `Browserify`, `EsLint`, `TypeScript`, `tslint`, `rollup` (*ES6 and ES5*), `tslib`, `espree`, `jquery`, `rx`, `ts-node`, `underscore`  and loads of `node_module` dependencies plus your libraries. And I also parsed `Esprima` and `Acorn` just for fun. Anyway. A few minor bugs found and was corrected. Uploaded latest to NPM.

**Comment by WebReflection** [comment]
Hi @KFlash, thanks a lot for your effort.

Just to be clear, I don't usually write ES2015+ code because I don't like transpilers and the issues these might cause.

However, it looks like the time to write modern code, now that every browser supports modules, has come.

Node is the last piece of the ecosystem that needs to work with them, and everything they are doing with this `.mjs` choice breaks somewhere else, hence my need for _cherow_ to work reliably with any sort of modern JS, and I'm enabling all jsx, v8, and next flags, and let me redefine import/exports.

majinbuu would be my first universal module (browsers and node) to use ascjs project so thanks again for making cherow more robust/battle tested.

**Comment by KFlash** [comment]
Hi. I'm are from the old school and not used to open source and this kind
of concepts. But after one of my clients used code from GH that never got
updated, lack of responses etc - Cherow become the end-result.

So because I know how frustrating it is to be in the middle of a
development process with a buggy dependency I try my best to make Cherow
what it should be.

As stated in the readme. You can read that I promise to always respond
quickly and solve bugs within 30 minutes.

So let's hope you succeed with your ongoing project and feel free to notify
me again if you find another bug.

**Comment by WebReflection** [comment]
btw .. `dist/cherow.js` fails to be parsed too :smile:

**Comment by WebReflection** [comment]
https://github.com/cherow/cherow/blob/master/dist/cherow.js#L169

```
/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:449
    throw createError.apply(void 0, [ type, this.getLocations() ].concat( params ));
    ^
Error: Line 169: Unexpected strict mode reserved word
    at constructError (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:306:17)
    at createError (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:327:17)
    at Parser.error (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:449:23)
    at Parser.isIdentifier (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:1862:24)
    at Parser.parseObjectElement (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:4083:14)
    at Parser.parseObjectExpression (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:4065:36)
    at Parser.parsePrimaryExpression (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:3784:25)
    at Parser.parseMemberExpression (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:3363:44)
    at Parser.parseLeftHandSideExpression (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:3354:25)
    at Parser.parseUpdateExpression (/home/webreflection/code/asbundle/node_modules/cherow/dist/cherow.js:3269:17)
```

**Comment by KFlash** [comment]
@WebReflection I will fix this shortly. That was the only library I didnt check, but easy fix !

Just now I ran into a nasty issue involving Acorn. Acorn refuse to parse this edge case `function *a({yield: b}){}`

**Comment by KFlash** [comment]
@WebReflection Fixed and updated to NPM. Now also parses the edge above, that Acorn doesn't.

**Comment by KFlash** [comment]
If there exist other bugs now, please open another issue ticket for them and I will fix it when I'm back online. Need to get my daughter now, and do some RL things.

**Comment by KFlash** [comment]
@WebReflection Need to mention one importing thing. This one `export { as as as  };`.  Esprima and Acorn does parse this. Shift does not. And if you try to parse it in NodeJS or V8 it fails too. This because the exported binding isn't declared. However. If you see [TC39](https://github.com/tc39/test262/blob/master/test/language/module-code/instn-star-id-name.js#L33) it should parse!! 
For now I choose to throw on this one.

**Comment by WebReflection** [comment]
for the time being it seems like a very weird edge case but I wonder why such choice ... I mean, you should not chose to diverge from standards or you'll miss the primary goal of this project, right?

**Comment by KFlash** [comment]
I did the choice in the first place because I started out with the idea of following V8, and as far as I know the V8 engine is almost 100% close to the specs. And the authors of Shift I guess you know who they are? So when both V8 and Shift choose to throw on it, and I already found plenty of bugs in the TC39 tests -  I did the same. However. I now followed Esprima and Acorn so It passes.
I just pushed a test for it.

**Comment by WebReflection** [comment]
I don’t know who Shift’s authors are but I’m pro specifications.
V8 might not be up to 100% compliance but it doesn’t mean it won’t try to.
Thanks for the update.

Unrelated question: is there a way to intercept all items while parsing, as opposite of me traversing the sub tree on a for/in?

Esprima apparently does that.

**Comment by KFlash** [comment]
Not at the moment, and I wasn't aware that Esprima can do that. I have more or less followed Acorn because it's more true to the standards. But there should be no problem to implement this soon as I get time for it. Do you have a link related to this or any documentation?

**Comment by WebReflection** [comment]
I’m waiting for Ariya to clarify on this, he told me so but I’m not sure myself.

The thing is that while traversing internally I could benefit from that effort instead of re-traversing all subnodes in a second step.

For my very specific use case that might be the best/fastest approach

**Comment by WebReflection** [comment]
I think it's this one: http://esprima.readthedocs.io/en/4.0/syntactic-analysis.html#syntax-delegate

**Comment by KFlash** [comment]
Ok. Same concept as Acorn's `onToken`. As you can see if you google Acorn's solution is the fastest approach - and same way as I collect comments. You can read [this](http://nick.balestra.ch/2015/pick-your-parsing/) old article. However. It shouldn't be too hard to implement this, the only issue is to adjust it correctly. I had this solution to begin with and you find trace of it [here](https://github.com/cherow/cherow/blob/master/src/masks.ts#L65). ATM. I got a terrible flu ATM, and are using my last resources before I collapse to make Cherow stable. If you want to take a stab at it, it should only take 10 - 15 min. to fix. In `nextToken` after scanned, just send the token found to a function.  More or less copy & paste comment collecting. However. You would need an sub-function to transform the token into correct names.

**Comment by KFlash** [comment]
I'm offline now. No more resources. But latest is on NPM. Hopefully stable now.

**Comment by KFlash** [comment]
@WebReflection I noticed that once again you claim that Cherow isn't stable? What exactly are you putting into the word "stable". Parsing everything after the "book" or have the features you need? 
I become curious.

First off I have to say I don't like to talk about others work, but I make an exception now. 
If we first look into the known parsers Acorn vs Esprima,  Acorn is the winner when it comes to spec compability. And in your special use case with module code a simple code like this `export {with as a}` and Esprima fails. Or `this export let[a] = 0 export let[b] = 0` where Esprima claims about reserved words, but in fact it should throw on duplicate `export`.  The list is longer.

Same thing for `Import`. Acorn parse correctly, Esprima fails.

So in this case, is this stable or unstable for you? I would say it's unstable. Anyway. Cherow passes too, and does the same thing as Acorn.

To take it further. In general code, you would also notice that Esprima fails multiple places on use of `yield` in binding context etc. Acorn doesn't. Is that stable or unstable?

Both Acorn and Esprima fails on Super property. Early error related. I would say it's something that hasn't been implemented for this parses. But for you, this is unstable? BTW. Cherow parses this too correctly.

To cut this down. And not to be rude. What do you mean with not stable this time?

**Comment by WebReflection** [comment]
Wow ... to start with, I'd like to be able to chose any dependency I want without needing to justify myself ... it's not like I've tweeted or published anywhere Cherow is not stable or anything, it was my own PR commit to myself so ... maybe ... chill?

To answer your points though, I'll try to explain my decision.

### My metrics
Both _ascjs_ and _asbundle_ were born basically for one reason: every other bundler is too heavy and I'm living abroad on data-roaming for 3+ months and it sucks.

![js](https://user-images.githubusercontent.com/85749/31867543-05952064-b768-11e7-8561-5d9aea1a5975.png)

When it comes to final npm package size, Esprima is the smaller out there.

My solutions doesn't need real-time performance so while I appreciate Cherow speed, I might not need it more than other metrics, including code size.

On top of that, the fact Cherow doesn't offer Acorn/Esprima like functionality means that whatever performance gain I have on Cherow gets lost the moment I need to manualy re-traverse the whole structure to find out import/export operations.

Accordingly, until there is such feature similar to Esprima delegate or Acorn onToken option (but I prefer Esprima simplicity over Acorn here), Cherow is [slower on bigger projects than Esprima](https://twitter.com/AriyaHidayat/status/920819193750077440).

### About Reliability
When I've opened the issue here and also filed a PR to solve them, the code was relatively trivial.

Only after you started testing Cherow against more real-world code, beside mine that like I've said was pointless to test. And I've thanked you for that.

However, I've set greenkeeper on to track your changes because you've pushed many patches one after the other like crazy and I've felt like you were rushing a bit too much in doing so.

Regardless  I wanted to be noticed about all changes, not just patches and greenkeeper has been on fire these days until I've noticed this:

![screenshot from 2017-10-22 20-36-51](https://user-images.githubusercontent.com/85749/31867630-9f8ad032-b769-11e7-95ec-ef38d7681d38.png)

But it's only less than a day ago that you pushed this too:
https://github.com/cherow/cherow/commit/c0732418908be68d9a8e18313d9b00227062c42d

That's not the way I bump versions on NPM though: I push changes that includes tests, I don't publish until tests ensure my new changes are working.

Accordingly, until I see that all tests are in place and you don't rush **publishing** on npm every line of code you change, I don't feel like this project can be considered stable, and I also have no time to update in sync both _ascjs_ and _asbundle_ every day ('cause I want to flat out dependency tree)

### About Parsing rules
Your argument for stable is based on parsing rules but this is not my meaning of stable when it comes to Open Source. If the best AST genrator ever bumps 3 versions in npm per day, that could still be the best AST generator, but many would not consider it stable.

However, and specially for my current needs, I don't want a binding proof parser, specially because _ascjs_ does not respect, transpile, understand, or want, ESM bindings at all because these are not simple to implement as CJS in a cross platform way, and surely these cannot land in _asbundle_ as fully cross browser solution, including old engines.

### As Summary
All I need these days is:
  * a robust parser with reasonable syntax tolerance that makes it easy for me to analyze **only** static ESM `import/export` and dynamic `require` in code
  * a parser that makes it possible to avoid me manually traversing the tree
  * a small as possible, with as few dependencies as possible, solution

I don't care about edge cases for now, I care about code stability.

I've also left both esprima and cherow branches up and merged esprima on master only because of greenkeeper (I don't know how to configure it for branches so ... I'm a very pragmatic developer).

So, what can I say, thanks a lot for your effort, I hope you'll finalize all the missing bits and offer a delegate function soon, and once you've done that I might re-consider using Cherow instead of any other.

But I have to be honest with you ... Acorn and it's ability to skip the whole tree is the best solution for me, because all I need when it comes to ESM is top level `import/export` and nothing else.

If Cherow would consider this possibility too, that'd be my best pick.

Best Regards

**Comment by KFlash** [comment]
Thank you for your explanation. And I'm chill. I just noticed your PR comment mention not stable, so that's the reason why I did this comment. Just to clarify.

Eprima on NPM takes around 283 KB and Acorn 117 KB (*I just checked*) 

Regarding the pushed to the repo. None of this changes have anything to do with NPM. Only one version have been published to NPM and rest to the repo.
But to be honest, we are more than one person using this user account and what I noticed now that the other one have just "maintained" - without pushing to NPM - while I was out of radio range and switched off due to a serious flu.

So to your needs. I guess your living conditions are similar to the country I live and work in just now, with poor internet connection and where each MB uploaded is almost worth more than gold. And people literally kill each other to become the first man to use mobile data.

I will address your concerns. I may need 1 - 2 days more to recover from my illness and than I will look into things again.

**Comment by WebReflection** [comment]
> Eprima on NPM takes around 283 KB and Acorn 117 KB (I just checked)

if you install via npm on an empty folder:

  * cherow is 558,8 kB
  * acorn is 440,6 kB
  * esprima is 328,9 kB

> Only one version have been published to NPM and rest to the repo.

I've received various greenkeeper PRs with like 4 npm published patches one after each other
https://github.com/WebReflection/ascjs/pull/4

> I will address your concerns. I may need 1 - 2 days more to recover from my illness and than I will look into things again.

get well soon

**Comment by KFlash** [comment]
That was chocking news! In all places I have seen Acorn is the smallest
one, and even Acorn claim it itself!? Wow. Good to know. Because I'm
reducing the Cherow code soon.

Anyway. I'm now on my way to collapse again, but I take other persons need
seriously. Attached is the evidence that what you need is almost working
now.

I will get back to it later on.

On Mon, Oct 23, 2017 at 8:28 AM, Andrea Giammarchi <notifications@github.com

**Comment by KFlash** [comment]
​

**Comment by KFlash** [comment]
@WebReflection Works only with functions for now, and I need to add the array for conversion. Here is latest: https://github.com/cherow/cherow/pull/46

**Comment by KFlash** [comment]
Hi. Sorry to say this. We changed our main idea / goal with Cherow and
choose to go for a 100% spec compat parser. As you can see from the latest
2 commits. This include we are not implementing any new features, only
accepting PR at this time. So for now. I wish you good luck with your
Esprima :)

**Comment by WebReflection** [comment]
the current goal doesn't logically exclude delegate/onToken capability but glad you told me so.

I'm removing Cherow branch from my projects, good luck with the project.

**Comment by KFlash** [comment]
I choose to focus on be 100% spec compatible and zero bugs. But I noticed
the last weeks that Open Source people doesn''t seem to care about
performance, spec compability or even if it's bug free!! So yes, I have to
disaspoint a lot of people from now on!

And for delegate/onToken, I didn't exclude it, if someone want to make it.
Let them make it and I will adopt that.

Good luck with your projects too!

On Mon, Oct 23, 2017 at 7:28 PM, Andrea Giammarchi <notifications@github.com

**Comment by WebReflection** [comment]
> I noticed the last weeks that Open Source people doesn''t seem to care about performance, spec compability or even if it's bug free

if you are talking about me you clearly have no idea who you are talking to + I've explicitly said what were my very specific needs for the projects that used Cherow

my specific needs != don't care about perf/spec/compatibility

you should _really_ chill man, I've already said on twitter cherow is great and you were very friendly too but now I regret doing so.

Don't push it too hard

**Comment by KFlash** [comment]
I think you misunderstood me. For the second. We only have written for a
few days - not weeks.  So this was not about you, but in general. What I
have experienced with open source after study their issue tracker.

When it comes to you... Well... I never talk about other projects or other
persons. But once more. What I noticed with you from day one - seen from my
the experience I have - is that you are very delegated to the task you are
doing. You set a goal, aim for it and give everything to reach it. A good
instinct! That was why I spent some extra time trying to fix things.
Also based on the way you communicate with me - the adult language you are
using - proves you are skilled too. Around same levels I guess.

So this is not about you and your skills. What I meant was clearly all
about what I can read in several issue trackers on GH. Everything from crap
language, to a demanding tone to fix that and that and so on. Like this
kind of
users take things for granted. That is not you, but what I have seen, read
and experienced the last weeks.

Hope that clearify things.

Again.!  Wish you good luck with your future projects!

On Mon, Oct 23, 2017 at 7:46 PM, Andrea Giammarchi <notifications@github.com

**Comment by WebReflection** [comment]
oh well, we agree then. I'm maintaining dozens projects myself here and thanks gosh not everyone is like that.

Take care.

**Comment by KFlash** [comment]
BTW. I'm going to add the tokenization stuff / onToken. No PR can fix it
correctly, so I have too. There exist loads of issues. The biggest issue is
that Esprima and Acorn tokenization are both flaws or have huge issues -
admitted by Acorn [here](https://github.com/ternjs/acorn/issues/589). On
the Esprima tracker you will find comments like "by design" as an excuse
not to fix that issue. And there are loads of other problems popping up
here.

I have never done this kind of coding before that would be needed to fix
this. So ......

On Mon, Oct 23, 2017 at 7:58 PM, Andrea Giammarchi <notifications@github.com

**Comment by WebReflection** [comment]
playing my "_for my current needs_" card: Esprima works perfectly fine for static top level `import/export` and dynamic `require` which is currently the only thing I'm interested in.

I might even end up writing an ASTree parser myself that does only that, if necessary, 'cause that's the only messed up part between Node and ES2015+ , the only part I'm trying to fix for my projects.

Regards

**Comment by KFlash** [comment]
Glad to hear you solved it your way :) Than I just dump my nasty looking
code that adds 230 lines of code bloat to Cherow source and no nothing than
tokenization. I probably implemented this the wrong way too.

On Mon, Oct 23, 2017 at 8:21 PM, Andrea Giammarchi <notifications@github.com

**Comment by KFlash** [comment]
I obviously was wrong thinking and mixed tokenization with delegation.
Anyway. I'm done. Cherow supports delegates now.
